### DISCUSSION TITLE
local_model의 .zero_grad() 필요!!

### DISCUSSION THREAD

**Discussion by right-chan** [N1-314/Intrinsic-Curiosity-Module---PyTorch#16] [General]
## 요약
- `local_model.zero_grad()` 필요
- 학습 성공

## 본문
- 저는 모델 파라미터를 다음과 같이 업데이트 했었습니다.
```python
for update in count():
    local_model.load_state_dict(global_model.state_dict())

    # rollout

    # calculate loss

    # parameter update
    optimizer.zero_grad()
    total_loss.backward()
    for local_param, global_param in zip(local_model.parameters(), global_model.parameters()):
        global_param.grad = local_param.grad.to(args.global_device)
    optimizer.step()
```
- loss가 1e23을 넘고 학습이 전혀 진행되지 않았습니다.
- RNN계열의 exploding gradient문제로 생각해서 `nn.utils.clip_grad_norm_(global_model.parameters(), max_norm=5)`gradient clip을 진행했지만 상황은 그대로였습니다.

Q: 무엇이 문제였을까요?
### A: local_model의 grad가 초기화되는 부분이 없습니다!

- 제가 사용한 optimizer는 `optimizer = SharedAdam(global_model.parameters(), lr=args.lr)`으로 `optimizer.zero_grad()`는 global_model 파라미터의 grad만 초기화 합니다.
- 따라서 `local_param.grad`는 계속해서 누적되고 있었습니다!

- gradient를 계산하는 모델은 local_model인데, global_model 파라미터만 reset gradient를 진행했습니다.
- `global_param.grad`는 `local_param.grad`로 넘겨받기에 `optimizer.zero_grad()`는 무용지물입니다.

### 여기서 필요한 것은?
**`local_model.zero_grad()`**

```python
for update in count():
    local_model.load_state_dict(global_model.state_dict())

    # rollout

    # calculate loss

    # parameter update
    optimizer.zero_grad()
    local_model.zero_grad() # add this
    total_loss.backward()
    for local_param, global_param in zip(local_model.parameters(), global_model.parameters()):
        global_param.grad = local_param.grad.to(args.global_device)
    optimizer.step()
```
- `local_model.zero_grad()`을 추가하여 `local_param.grad`를 초기화
- 우리의 의도대로 동작 -> **학습 성공**
- 추가로 `optimizer.zero_grad()`는 없어도 동작할 것 같습니다만 빼고 실험은 안해봤습니다.

**Comment by HyeAnn-Lee** [comment]
[Zeroing out gradients in PyTorch](https://pytorch.org/tutorials/recipes/recipes/zeroing_out_gradients.html#zero-the-gradients-while-training-the-network)

> You can also use `model.zero_grad()`. This is the same as using `optimizer.zero_grad()` as long as all your model parameters are in that optimizer. Use your best judgment to decide which one to use.

**Comment by right-chan** [comment]
```python
for update in count():
    local_model.load_state_dict(global_model.state_dict(), assign=True) # add this

    # rollout

    # calculate loss

    # parameter update
    optimizer.zero_grad()
    # local_model.zero_grad() # add this
    total_loss.backward()
    for local_param, global_param in zip(local_model.parameters(), global_model.parameters()):
        global_param.grad = local_param.grad.to(args.global_device)
    optimizer.step()
```

`load_state_dict`에서 `assign`을 활성화하면 `local_model.zero_grad()`가 없어도 `local_model.weight.grad`가 초기화 됩니다.

https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.load_state_dict

**Reply by right-chan** [reply]
그런데 정확한 사용법인지는 모르겠어요..
